tempesta_fw/http.c

```diff
@@ -3383,8 +3384,8 @@ tfw_http_req_process(TfwConn *conn, TfwStream *stream, const TfwFsmData *data)
 	 * feed the new data version to FSMs registered on our states.
 	 */
 	data_up.skb = skb;
-	data_up.off = off;
-	data_up.trail = curr_skb_trail;
+	data_up.off = 0;
```

Comment:
Yes, agree with @ikoveshnikov on both points.

Reply:
That's confusing.
In the code below there are calls of `tfw_http_chop_skb` before calls of `tfw_gfsm_move()`. The one in the `TFW_POSTPONE` case is what I recently added, but the one in the `TFW_PASS` (falls through the switch-case) was there before. So `TfwFsmData` instance was passed with `off` set to possibly non-zero value. But at the same time, `tfw_http_chop_skb()` ensured that that `off` bytes were cut off.

So... what `data_up.off` should actually be? Is it `0` or `off`?

Similarly, `data_up.trail = curr_skb_trail;` confuses me. As `curr_skb_trail` can be either `trail` (for the last message in skb) or `0`, `data_up` got even the trailing data which it shouldn't be getting. But it all worked somehow. I wonder how.

I was hoping to avoid digging further into the overall data processing. But it looks inevitable now. :confused:

Reply:
>But it all worked somehow. I wonder how.

It seems, that was a hidden bug. Perhaps it is worth to move `tfw_http_chop_skb()` before `data_up` setting (like in `tfw_tls_msg_process()` or in `tfw_h2_frame_process()`) - to deal with offsets and trails from the previous layer at first, and after that - prepare `data_up` for further transferring upstairs. In this variant `data_up.off` really should be 0 here, and `data_up.len` should be set to `parsed`.

Reply:
>  Perhaps it is worth to move tfw_http_chop_skb() before data_up setting 

Sounds like a good idea. All that offsets and lengths create an easy way to get lost and garbage data.

> But it all worked somehow. I wonder how.

Actually it's OK for now to have invalid `off` and `len` members values in `data_up`. Since there is no code that uses them (: The only FSM called after HTTP is Frang (http_limits), it doesn't work with that members. Even if there is issues, we don't see them.

The issue with incorrect trail in `tfw_http_chop_skb()` can be also hidden by `ss_skb_split()` which is called right after the `tfw_http_chop_skb()`. Not sure this really happens, but it also cat chop extra data after request end...

Reply:
Yes, there is some mess with message (HTTP requests and TLS application data records) chopping and the code redesign is required.

`tfw_http_chop_skb()` and `tfw_tls_chop_skb_rec()` seems do very close things, but do them on different layers. Probably, we need only one of them and/or can unify the functions. I'd suppose that chopping must be called by TLS layer only since it has knowledge about TLS records. `tfw_http_chop_skb()` just uses the knowledge passed from TLS layer.

Initially `TfwFsmData` `off` and `trail` (now `len`) were introduced to make HTTP and TLS layers independent: TLS knows about record header and trailer and calls HTTP FSM with explicit description which data must be ignored. If we chop data, then why do we need `off` and `len`? The things are even more simple with that we assemble full TLS record and know its header and trailer before decryption and processing on HTTP level. Probably, with proper chopping, we can simplify `TfwFsmData` structure.

Reply:
> Perhaps it is worth to move tfw_http_chop_skb() before data_up setting

Moved it to the earlier lines.

Reply:
There were changes that removed `off` and `trail`. They are no longer cut here, and instead are cut in the TLS layer.